examples/plot_metrics_with_error.py

```diff
@@ -0,0 +1,163 @@
+from fairlearn.metrics import MetricFrame, ErrorPlotter
+
+
+data = fetch_openml(data_id=1590, as_frame=True)
+X = pd.get_dummies(data.data)
+y_true = (data.target == '>50K') * 1
+sex = data.data['sex']
+
+# %%
+# Error Metrics
+# -------------
+# We have many different choices for error metrics. In this notebook we'll just be using a
+# Normal approximation interval (symmetric) and a Wilson score interval (asymmetric).
+#
+# https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Wilson_score_interval
+# https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Normal_approximation_interval
+#
+#
+
+z_score = 1.959964
+digits = 4
+
+
+def recall_normal_err(y_t, y_p):
+    assert len(y_t) == len(y_p)
+    tn, fp, fn, tp = confusion_matrix(y_t, y_p).ravel()
+    error = compute_error_metric(tp/(tp+fn), tp + fn, z_score=z_score)
+    return error
+
+
+def wilson(p, n, digits=digits, z=z_score):
+    """ Returns lower and upper bound """
+    denominator = 1 + z**2/n
+    centre_adjusted_probability = p + z*z / (2*n)
+    adjusted_standard_deviation = np.sqrt((p*(1 - p) + z*z / (4*n)))/np.sqrt(n)
+    lower_bound = (centre_adjusted_probability - z*adjusted_standard_deviation) / denominator
+    upper_bound = (centre_adjusted_probability + z*adjusted_standard_deviation) / denominator
+    return (round(lower_bound, digits), round(upper_bound, digits))
+
+
+def compute_error_metric(metric_value, sample_size, z_score):
+    return z_score*np.sqrt(metric_value*(1.0-metric_value))/np.sqrt(sample_size)
+
+
+def recall_wilson_lower_bound(y_t, y_p):
+    assert len(y_t) == len(y_p)
+    tn, fp, fn, tp = confusion_matrix(y_t, y_p).ravel()
+    bounds = wilson(tp/(tp+fn), tp + fn, digits, z_score)
+    return bounds[0]
+
+
+def recall_wilson_upper_bound(y_t, y_p):
+    assert len(y_t) == len(y_p)
+    tn, fp, fn, tp = confusion_matrix(y_t, y_p).ravel()
+    bounds = wilson(tp/(tp+fn), tp + fn, digits, z_score)
+    return bounds[1]
+
+
+# %%
+# Classifier
+# ----------
+# Now we use a Decision Tree to make predictions
+#
+
+classifier = DecisionTreeClassifier(min_samples_leaf=10, max_depth=4)
+classifier.fit(X, y_true)
+y_pred = classifier.predict(X)
+
+# %%
+# MetricFrame
+# -----------
+# Now we create a MetricFrame to generate the Recall wilson bounds and Recall normal error
+#
+
+# Analyze metrics using MetricFrame
+metrics = {
+    'Recall': recall_score,
+    'Recall lower bound': recall_wilson_lower_bound,
+    'Recall upper bound': recall_wilson_upper_bound,
+    'Recall Error': recall_normal_err
+}
+metric_frame = MetricFrame(metrics, y_true, y_pred, sensitive_features=sex)
+
+# %%
+# Error Plotting
+# --------------
+# Finally we use the ErrorPlotter class to plot the Recall metric with its error bounds
+#
+# 1. We plot with the symmetric Normal Error Interval
+error_mapping = {
+    "Recall": {
+        "symmetric_error": "Recall Error",
+    }
+}
+
+ep = ErrorPlotter(metric_frame, error_mapping)
```

Comment:
I would encourage feedback from data scientists on this API... @hildeweerts  @MiroDudik @michaelamoako

Reply:
I kind of see 3 options:

## OPTION 1 - the most freedom + work
MetricFrame contains metrics and metric errors, where mapping is given by error_mapping
- User also has freedom of naming metric errors

```
metrics = {
    'Recall': recall_score,
    'Recall Error': recall_normal_err
}
metric_frame = MetricFrame(metrics, y_true, y_pred, sensitive_features=sex)

error_mapping = {
    "Recall": {                             # `Recall` is the Metric Name
        "symmetric_error": "Recall Error",  # `symmetric_error` is a predefined type of error metric
                                            # `Recall Error` is the column name of the metric defined in the MetricFrame
    }
}

ep = ErrorPlotter(metric_frame, error_mapping)
ep.plot_with_error("bar", metric="Recall",
                   title="(Symmetric) Normal Error Intervals", capsize=10, colormap="Pastel1")
```

## OPTION 2 - middle freedom/work

MetricFrame only has metrics, not metric errors; error_mapping contains the error_metric-generating function
-  Fairlearn would calculate and/or append metric errors to MetricFrame (user might not have control over the name of generated metric errors?)

```
metrics = {
    'Recall': recall_score,
}
metric_frame = MetricFrame(metrics, y_true, y_pred, sensitive_features=sex)

error_mapping = {
    "Recall": {                                # `Recall` is the Metric Name
        "symmetric_error": recall_normal_err,  # `symmetric_error` is a predefined type of error metric
                                               # `recall_normal_err` is a function that generates the error metric
    }
}

ep = ErrorPlotter(metric_frame, error_mapping)
ep.plot_with_error("bar", metric="Recall",
                   title="(Symmetric) Normal Error Intervals", capsize=10, colormap="Pastel1")
```
## OPTION 3 - least freedom/work

MetricFrame contains the metrics, metric errors, and the mapping
-  Force a specific appended string for each symmetric_error/lower_bound/upper_bound/etc i.e. `Recall_symmetric_error`
- Explicitly make the user follow our given string mapping

```
metrics = {
    'Recall': recall_score,
    'Recall_symmetric_error': recall_normal_err
}
metric_frame = MetricFrame(metrics, y_true, y_pred, sensitive_features=sex)

ep = ErrorPlotter(metric_frame)
ep.plot_with_error("bar", metric="Recall",
                   title="(Symmetric) Normal Error Intervals", capsize=10, colormap="Pastel1")
```

Reply:
Additionally, it might be better to make ErrorPlotter.plot_with_error() a static function, since I'm not exactly sure what a benefit of creating different instances may be

Reply:
@hildeweerts  in your example, what would be the arguments of the `errorfunction`? Wouldn't the errors need to be calculated with access to all the data, which are only available at the point of `MetricFrame` construction?

Reply:
@riedgar-ms I'm not sure if I understand your question - I was under the impression that in the `error_mapping` you could choose between either a predefined string or pass a function that will be used to compute the errors. Kind of like how in scikit-learn you can often pass either predefined string `'accuracy'` or the function `accuracy_score`. But perhaps I misunderstood?

Reply:
For the purposes of an internship, I think the last bullet point is the one to use. Looking at it implemented, I agree that there's nothing which really needs to be an `ErrorPlotter` object (I don't see internal state which would justify it).

Since we don't have a way for the `MetricFrame` to record the mapping for now (I'm really not a fan of encoding that information in the column names), that will have to remain a separate dictionary as @alexquach has here.

Reply:
Before `MetricFrame` supports this natively, I think I'd prefer a more "shallow" API. I find the dictionary too complicated for my taste. How about something like:
```python
metrics = {
    'Recall': recall_score,
    'Recall Error': recall_normal_err,
    'Recall Conf Int': recall_wilson,
    'FPR': false_positive_rate,
    'FPR Error': false_positive_rate_err,
    'FPR Conf Int': false_positive_rate_wilson,
}
mf = MetricFrame(metrics=metrics, y_true=y_true, y_pred=y_pred, sensitive_features=sf)

# plot metrics with (symmetric) error bars
plot_metric_frame(mf, metrics=['Recall', 'FPR'], error_bars=['Recall Error', 'FPR Error'])
plot_metric_frame(mf, metrics='Recall', error_bars='Recall Error')

# plot metrics with confidence intervals
plot_metric_frame(mf, metrics=['Recall', 'FPR'], conf_intervals=['Recall Conf Int', 'FPR Conf Int'])
plot_metric_frame(mf, metrics='Recall', conf_intervals='Recall Error')

# plot metrics without error bars
plot_metric_frame(mf, metrics=['Recall', 'FPR'])

# plots all columns and treats them as metrics without error bars
plot_metric_frame(mf) 
```

Reply:
@MiroDudik Yeah I think we can simplify / make a shallower API by requiring the user to provide bounds/errors (and therefore their metric-generating function) as a tuple rather than allowing lower_bounds+upper_bounds / lower_error + upper_error.

Reply:
@alexquach : Not sure what you mean. What do you think about my proposal above? It shows you how to specify either scalar errors (these would be plotted symmetric around the metric value) or intervals (these would be returned as pairs, similar to `wilson` in your code).

Reply:
@MiroDudik are you OK with only allowing confidence intervals *or* error bars in a single plot? That is one consequence of using lists rather than a `dict`.

Reply:
Yes. That's intentional. There's a trivial work-around (via representing error-bars as confidence intervals), so I'd just go with this API for the sake of clarity.